File: src/main/java/net/dunciboy/discord_bot/commands/BanUserById.java
/*
 * Copyright 2017 Duncan C.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package net.dunciboy.discord_bot.commands;

import net.dunciboy.discord_bot.GuildLogger;
import net.dunciboy.discord_bot.JDALibHelper;
import net.dunciboy.discord_bot.RunBots;
import net.dv8tion.jda.core.EmbedBuilder;
import net.dv8tion.jda.core.MessageBuilder;
import net.dv8tion.jda.core.Permission;
import net.dv8tion.jda.core.entities.ChannelType;
import net.dv8tion.jda.core.entities.Member;
import net.dv8tion.jda.core.entities.Message;
import net.dv8tion.jda.core.entities.PrivateChannel;
import net.dv8tion.jda.core.events.message.MessageReceivedEvent;

import java.awt.*;
import java.util.concurrent.TimeUnit;

/**
 * Created by Duncan on 22/05/2017.
 * This class creates a command to ban users by id with logging.
 */
public class BanUserById extends Ban {
    private static final String[] ALIASES = new String[]{"BanByUserId", "BanById"};
    private static final String ARGUMENTATION_SYNTAX = "[user id] [reason~]";
    private static final String DESCRIPTION = "Will ban the user with the id, clear all message that where posted by the user the last 24 hours and log it to the log channel.";


    /**
     * Constructor for abstract class
     */
    public BanUserById() {
        super(ALIASES, ARGUMENTATION_SYNTAX, DESCRIPTION);
    }

    @Override
    void commandExec(MessageReceivedEvent event, String arguments, PrivateChannel privateChannel) {
        if (!event.isFromType(ChannelType.TEXT)) {
            if (privateChannel != null) {
                event.getChannel().sendMessage("This command only works in a guild.").queue(message -> message.delete().queueAfter(1, TimeUnit.MINUTES));
            }
        } else if (!event.getMember().hasPermission(Permission.BAN_MEMBERS)) {
            event.getChannel().sendMessage("You need ban members permission to use this command.").queue(message -> message.delete().queueAfter(1, TimeUnit.MINUTES));
        } else {
            String userId;
            try {
                userId = arguments.split(" ")[0];
            } catch (NullPointerException e) {
                throw new IllegalArgumentException("No id provided");
            }
            String reason;
            try {
                reason = arguments.substring(arguments.split(" ")[0].length() + 1);
            } catch (IndexOutOfBoundsException e) {
                throw new IllegalArgumentException("No reason provided for this action.");
            }
            event.getJDA().retrieveUserById(userId).queue(toBan ->
            {
                Member toBanMemberCheck = event.getGuild().getMember(toBan);
                if (toBan != null && !event.getMember().canInteract(toBanMemberCheck)) {
                    if (privateChannel != null) {
                        privateChannel.sendMessage("You can't ban a user that you can't interact with.").queue();
                    }
                    return;
                }
                event.getGuild().getController().ban(toBan, 1, reason).queue(aVoid -> {
                    RunBots runBots = RunBots.Companion.getRunBot(event.getJDA());
                    if (runBots != null) {
                        EmbedBuilder logEmbed = new EmbedBuilder()
                                .setColor(Color.RED)
                                .setTitle("User was banned by id  | Case: " + GuildLogger.getCaseNumberSerializable(event.getGuild().getIdLong()))
                                .addField("User", toBan.getName(), true)
                                .addField("Moderator", JDALibHelper.INSTANCE.getEffectiveNameAndUsername(event.getMember()), true)
                                .addField("Reason", reason, false);

                        runBots.getLogToChannel().log(logEmbed, toBan, event.getGuild(), null);
                    }

                    if (privateChannel == null) {
                        return;
                    }

                    Message creatorMessage = new MessageBuilder()
                            .append("Banned ").append(toBan.toString())
                            .build();
                    privateChannel.sendMessage(creatorMessage).queue();
                }, throwable -> {
                    if (privateChannel == null) {
                        return;
                    }

                    Message creatorMessage = new MessageBuilder()
                            .append("Banning user failed\n")
                            .append(throwable.getClass().getSimpleName()).append(": ").append(throwable.getMessage())
                            .build();
                    privateChannel.sendMessage(creatorMessage).queue();
                });
            }, throwable -> {
                if (privateChannel == null) {
                    return;
                }

                Message creatorMessage = new MessageBuilder()
                        .append("Failed retrieving the user, banning failed.\n")
                        .append(throwable.getClass().getSimpleName()).append(": ").append(throwable.getMessage())
                        .build();
                privateChannel.sendMessage(creatorMessage).queue();
            });
        }
    }
}

File: src/main/kotlin/be/duncanc/discordmodbot/bot/commands/BanUserById.kt
/*
 * Copyright 2018 Duncan Casteleyn
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package be.duncanc.discordmodbot.bot.commands

import be.duncanc.discordmodbot.bot.services.GuildLogger
import be.duncanc.discordmodbot.bot.utils.nicknameAndUsername
import net.dv8tion.jda.api.EmbedBuilder
import net.dv8tion.jda.api.MessageBuilder
import net.dv8tion.jda.api.Permission
import net.dv8tion.jda.api.entities.ChannelType
import net.dv8tion.jda.api.entities.PrivateChannel
import net.dv8tion.jda.api.events.message.MessageReceivedEvent
import org.springframework.stereotype.Component
import java.awt.Color
import java.util.*
import java.util.concurrent.TimeUnit


/**
 * Created by Duncan on 22/05/2017.
 * This class creates a command to ban users by id with logging.
 */
/**
 * Constructor for abstract class
 */
@Component
class BanUserById : CommandModule(
    arrayOf("BanByUserId", "BanById"),
    "[user id] [reason~]",
    "Will ban the user with the id, clear all message that where posted by the user the last 24 hours and log it to the log channel.",
    true,
    true
) {

    public override fun commandExec(event: MessageReceivedEvent, command: String, arguments: String?) {
        event.author.openPrivateChannel().queue(
            { privateChannel -> commandExec(event, arguments, privateChannel) }
        ) { commandExec(event, arguments, null as PrivateChannel?) }
    }

    fun commandExec(event: MessageReceivedEvent, arguments: String?, privateChannel: PrivateChannel?) {
        if (!event.isFromType(ChannelType.TEXT)) {
            if (privateChannel != null) {
                event.channel.sendMessage("This command only works in a guild.")
                    .queue { message -> message.delete().queueAfter(1, TimeUnit.MINUTES) }
            }
        } else if (event.member?.hasPermission(Permission.BAN_MEMBERS) != true) {
            event.channel.sendMessage("You need ban members permission to use this command.")
                .queue { message -> message.delete().queueAfter(1, TimeUnit.MINUTES) }
        } else {
            val userId: String
            try {
                userId = arguments!!.split(" ".toRegex()).dropLastWhile { it.isEmpty }.toTypedArray()[0]
            } catch (e: NullPointerException) {
                throw IllegalArgumentException("No id provided")
            }

            val reason: String
            try {
                reason =
                    arguments.substring(arguments.split(" ".toRegex()).dropLastWhile { it.isEmpty }
                        .toTypedArray()[0].length + 1)
            } catch (e: IndexOutOfBoundsException) {
                throw IllegalArgumentException("No reason provided for this action.")
            }

            event.jda.retrieveUserById(userId).queue({ toBan ->
                val toBanMemberCheck = event.guild.getMember(toBan)
                if (toBanMemberCheck != null && event.member?.canInteract(toBanMemberCheck) != true) {
                    privateChannel?.sendMessage("You can't ban a user that you can't interact with.")?.queue()
                    return@queue
                }
                event.guild.ban(userId, 1).queue(banQue@{
                    val guildLogger = event.jda.registeredListeners.firstOrNull { it is GuildLogger } as GuildLogger?
                    if (guildLogger != null) {
                        val logEmbed = EmbedBuilder()
                            .setColor(Color.RED)
                            .setTitle("User banned by id")
                            .addField("UUID", UUID.randomUUID().toString(), false)
                            .addField("User", toBan.name, true)
                            .addField("Moderator", event.member!!.nicknameAndUsername, true)
                            .addField("Reason", reason, false)

                        guildLogger.log(logEmbed, toBan, event.guild, null, GuildLogger.LogTypeAction.MODERATOR)
                    }

                    if (privateChannel == null) {
                        return@banQue
                    }

                    val creatorMessage = MessageBuilder()
                        .append("Banned ").append(toBan.toString())
                        .build()
                    privateChannel.sendMessage(creatorMessage).queue()
                }) banQueThrowable@{ throwable ->
                    if (privateChannel == null) {
                        return@banQueThrowable
                    }

                    val creatorMessage = MessageBuilder()
                        .append("Banning user failed\n")
                        .append(throwable.javaClass.simpleName).append(": ").append(throwable.message)
                        .build()
                    privateChannel.sendMessage(creatorMessage).queue()
                }
            }) { throwable ->
                if (privateChannel == null) {
                    return@queue
                }

                val creatorMessage = MessageBuilder()
                    .append("Failed retrieving the user, banning failed.\n")
                    .append(throwable.javaClass.simpleName).append(": ").append(throwable.message)
                    .build()
                privateChannel.sendMessage(creatorMessage).queue()
            }
        }
    }
}
